cirq/ops/n_qubit_diagonal_gate.py

```diff
@@ -0,0 +1,188 @@
+
+        This gate's off-diagonal elements are zero and it's on diagonal
+        elements are all phases.
+
+        Args:
+            diag_angles_radians: The list of angles on the diagonal in radians.
+                If these values are $(x_0, x_1, \ldots , x_N)$ then the unitary
+                has diagonal values $(e^{i x_0}, e^{i x_1}, \ldots, e^{i x_N})$.
+        """
+        self._diag_angles_radians: Tuple[value.TParamVal, ...] = tuple(diag_angles_radians)
+
+    def _num_qubits_(self):
+        return int(np.log2(len(self._diag_angles_radians)))
+
+    def _is_parameterized_(self) -> bool:
+        return any(protocols.is_parameterized(angle) for angle in self._diag_angles_radians)
+
+    def _parameter_names_(self) -> AbstractSet[str]:
+        return {
+            name for angle in self._diag_angles_radians for name in protocols.parameter_names(angle)
+        }
+
+    def _resolve_parameters_(
+        self, param_resolver: 'cirq.ParamResolver', recursive: bool
+    ) -> 'DiagonalGate':
+        return DiagonalGate(
+            protocols.resolve_parameters(self._diag_angles_radians, param_resolver, recursive)
+        )
+
+    def _has_unitary_(self) -> bool:
+        return not self._is_parameterized_()
+
+    def _unitary_(self) -> np.ndarray:
+        if self._is_parameterized_():
+            return None
+        return np.diag([np.exp(1j * angle) for angle in self._diag_angles_radians])
+
+    def _apply_unitary_(self, args: 'protocols.ApplyUnitaryArgs') -> np.ndarray:
+        if self._is_parameterized_():
+            return NotImplemented
+        for index, angle in enumerate(self._diag_angles_radians):
+            subspace_index = args.subspace_index(big_endian_bits_int=index)
+            args.target_tensor[subspace_index] *= np.exp(1j * angle)
+        return args.target_tensor
+
+    def _circuit_diagram_info_(
+        self, args: 'cirq.CircuitDiagramInfoArgs'
+    ) -> 'cirq.CircuitDiagramInfo':
+        rounded_angles = np.array(self._diag_angles_radians)
+        if args.precision is not None:
+            rounded_angles = rounded_angles.round(args.precision)
+        if len(rounded_angles) <= 4:
+            diag_str = 'diag({})'.format(', '.join(proper_repr(angle) for angle in rounded_angles))
+        else:
+            diag_str = ', '.join(proper_repr(angle) for angle in rounded_angles[:2])
+            diag_str += ', ..., '
+            diag_str += ', '.join(proper_repr(angle) for angle in rounded_angles[-2:])
+            diag_str = 'diag({})'.format(diag_str)
+        return protocols.CircuitDiagramInfo(
+            [diag_str] + ['#' + str(i) for i in range(2, self._num_qubits_() + 1)]
+        )
+
+    def __pow__(self, exponent: Any) -> 'DiagonalGate':
+        if not isinstance(exponent, (int, float, sympy.Basic)):
+            return NotImplemented
+        angles = []
+        for angle in self._diag_angles_radians:
+            mulAngle = protocols.mul(angle, exponent, NotImplemented)
```

Comment:
Prefer snake case over camel case
```suggestion
            mul_angle = protocols.mul(angle, exponent, NotImplemented)
```